File: src/ir.rs
use std::collections::BTreeMap;

use serde::Deserialize;

pub type Block = Vec<Exec>;

/// This struct stores immutable data such as code.
#[derive(Debug, Deserialize)]
pub struct Data {
   pub functions: Vec<Function>,
   pub prototypes: Vec<Prototype>,
   /// ident id -> ident name
   pub ident_map: BTreeMap<usize, String>,
   pub reserved_idents: ReservedIdents,
   pub files: Vec<String>
}

#[derive(Debug, Deserialize)]
pub struct Range {
   /// Index for Data.files
   pub file: usize,
   /// 0-indexed line number
   pub line: usize,
   /// 0-indexed column number
   pub col: usize,
   /// length of token
   pub len: usize
}

#[derive(Debug, Deserialize)]
pub struct ReservedIdents {
   pub next: usize,
   pub __call__: usize
}

#[derive(Debug, Deserialize)]
pub struct Prototype {
   pub name: String,
   /// field ident id -> index for Struct.values
   pub field_map: BTreeMap<usize, usize>,
   /// method ident id -> index for Data.functions
   pub method_map: BTreeMap<usize, usize>
}

#[derive(Debug, Deserialize)]
pub struct Function {
   pub name: String,
   pub parameters: Vec<String>,
   pub variables: Vec<String>,
   pub body: Block
}

/// Literal values for primitive data-types
#[derive(Debug, Deserialize)]
pub enum Literal {
   Nil,
   IterEnd,
   Bool(bool),
   Int(i64),
   Float(f64),
   Str(String)
}

/// Operators which take 1 parameter
#[derive(Debug, Deserialize)]
pub enum UnaryOperator {
   Not,
   BitNot,
   Minus,
   Type,
   Err,
   Bool,
   Int,
   Float,
   Str,
   Len,
   Print,
   Read
}

/// Operators which take 2 parameters
#[derive(Debug, Deserialize)]
pub enum BinaryOperator {
   Add,
   Sub,
   Mul,
   Div,
   Modulo,
   GetItem,
   Eq,
   Is,
   Lt,
   Leq,
   BitAnd,
   BitOr,
   BitXor,
   LeftShift,
   RightShift,
   And,
   Or,
   Push,
   Remove,
   Index,
   Join,
   Write
}

/// Operators which take 3 parameters
#[derive(Debug, Deserialize)]
pub enum TernaryOperator {
   Branch,
   SetItem
}

/// Operators which take any no. of parameters
#[derive(Debug, Deserialize)]
pub enum NaryOperator {
   List
}

/// Reference stores index to variables or functions
#[derive(Debug, Deserialize)]
pub enum Reference {
   /// Index offset to Data.variables
   Variable(usize),
   /// Index to Data.functions
   Function(usize)
}

#[derive(Debug, Deserialize)]
#[serde(tag = "type")]
pub enum Exec {
   While {
      condition: Expr,
      block: Block
   },
   ForLoop {
      variable: Reference,
      iterator: Expr,
      block: Block
   },
   DoWhile {
      block: Block,
      condition: Expr
   },
   Branch {
      condition: Expr,
      then: Block,
      otherwise: Block
   },
   Return {
      expr: Expr
   },
   /// Statement comprising of a expression whose evaluated value is immediately
   /// dropped.
   Expr {
      expr: Expr
   }
}

#[derive(Debug, Deserialize)]
#[serde(tag = "type")]
pub enum Expr {
   Literal { literal: Literal },
   Reference { reference: Reference },
   UnaryOperation { operator: UnaryOperator, expr: Box<Expr> },
   BinaryOperation { operator: BinaryOperator, left: Box<Expr>, right: Box<Expr> },
   TernaryOperation { operator: TernaryOperator, first: Box<Expr>, second: Box<Expr>, third: Box<Expr> },
   NaryOperation { operator: NaryOperator, parameters: Vec<Expr> },
   Call { callable: Box<Expr>, parameters: Vec<Expr> },
   Plugin { id: usize, parameters: Vec<Expr> },
   Struct { prototype: usize, values: Vec<Expr> },
   SetVar { variable: Reference, expr: Box<Expr> },
   SetField { instance: Box<Expr>, field_id: usize, value: Box<Expr> },
   GetField { instance: Box<Expr>, field_id: usize },
   Die { expr: Box<Expr>, range: Range },
   OrDie { expr: Box<Expr>, range: Range }
}

File: onyoc/ir.py
from dataclasses import dataclass
from enum import auto

from .serde import *


class ExecT:
   ...


class LiteralT:
   ...


Block = list[ExecT]


@dataclass
class Data(Struct_):
   functions: list["Function"]
   prototypes: list["Prototype"]
   ident_map: dict[int, str]
   reserved_idents: "ReservedIdents"


@dataclass
class ReservedIdents(Struct_):
   next: int

   @staticmethod
   def from_ident_map(ident_map: dict[str, int]):
      return ReservedIdents(next=ident_map["next"])


@dataclass
class Function(Struct_):
   name: str
   parameters: list[str]
   variables: list[str]
   body: Block


@dataclass
class Prototype(Struct_):
   name: str
   field_map: dict[int, int]
   method_map: dict[int, int]


class Literal(ExternallyTaggedEnum):
   @dataclass
   class Nil(LiteralT, ValueEnum):
      _: None = None

   @dataclass
   class Bool(LiteralT, ValueEnum):
      _: bool

   @dataclass
   class Int(LiteralT, ValueEnum):
      _: int

   @dataclass
   class Float(LiteralT, ValueEnum):
      _: float

   @dataclass
   class Str(LiteralT, ValueEnum):
      _: str


class UnaryOperator(Enum):
   Not = auto()
   BitNot = auto()
   Minus = auto()
   Type = auto()
   Err = auto()
   Bool = auto()
   Int = auto()
   Float = auto()
   Str = auto()
   Len = auto()
   Print = auto()
   Read = auto()


class BinaryOperator(Enum):
   Add = auto()
   Sub = auto()
   Mul = auto()
   Div = auto()
   Modulo = auto()
   GetItem = auto()
   Eq = auto()
   Is = auto()
   Lt = auto()
   Leq = auto()
   BitAnd = auto()
   BitOr = auto()
   BitXor = auto()
   LeftShift = auto()
   RightShift = auto()
   And = auto()
   Or = auto()
   Push = auto()
   Remove = auto()
   Index = auto()
   Join = auto()
   Write = auto()


class TernaryOperator(Enum):
   Branch = auto()
   SetItem = auto()


class NaryOperator(Enum):
   Tuple = auto()
   List = auto()


class ReferenceT:
   ...


class Reference(ExternallyTaggedEnum):
   @dataclass
   class Variable(ReferenceT, ValueEnum):
      _: int

   @dataclass
   class Function(ReferenceT, ValueEnum):
      _: int


class ExprT:
   ...


class Exec(InternallyTaggedEnum):
   @dataclass
   class While(ExecT, Struct_):
      condition: ExprT
      block: Block

   @dataclass
   class ForLoop(ExecT, Struct_):
      variable: ReferenceT
      iterator: ExprT
      block: Block

   @dataclass
   class DoWhile(ExecT, Struct_):
      block: Block
      condition: ExprT

   @dataclass
   class Branch(ExecT, Struct_):
      condition: ExprT
      then: Block
      otherwise: Block

   @dataclass
   class Return(ExecT, Struct_):
      expr: ExprT

   @dataclass
   class Expr(ExecT, Struct_):
      expr: ExprT


class Expr(InternallyTaggedEnum):
   @dataclass
   class Literal(ExprT, Struct_):
      literal: LiteralT

   @dataclass
   class Reference(ExprT, Struct_):
      reference: ReferenceT

   @dataclass
   class UnaryOperation(ExprT, Struct_):
      operator: UnaryOperator
      expr: ExprT

   @dataclass
   class BinaryOperation(ExprT, Struct_):
      operator: BinaryOperator
      left: ExprT
      right: ExprT

   @dataclass
   class TernaryOperation(ExprT, Struct_):
      operator: TernaryOperator
      first: ExprT
      second: ExprT
      third: ExprT

   @dataclass
   class NaryOperation(ExprT, Struct_):
      operator: NaryOperator
      parameters: list[ExprT]

   @dataclass
   class Call(ExprT, Struct_):
      callable: ExprT
      parameters: list[ExprT]

   @dataclass
   class Struct(ExprT, Struct_):
      prototype: int
      values: list[ExprT]

   @dataclass
   class SetVar(ExprT, Struct_):
      variable: ReferenceT
      expr: ExprT

   @dataclass
   class SetField(ExprT, Struct_):
      instance: ExprT
      field_id: int
      value: ExprT

   @dataclass
   class GetField(ExprT, Struct_):
      instance: ExprT
      field_id: int
